[!!!][TASK] Feature refactor

skipperbent opened the pull request.

skipperbent committed:
Added HttpRequest + HttpResponse objects

skipperbent committed:
[TASK] Renaming stuff to make more sense

skipperbent committed:
[TASK] Moved _data to RestBase so parameters can be easily extended

skipperbent committed:
[TASK] Began refactor SDK to work with other Restful API's and made it more user- and debugging friendly

skipperbent committed:
[CLEANUP] Minor cleanup

skipperbent committed:
[TASK] More cleanup, readded staging url for TRAPP service.

skipperbent commented:
Refs: #9 #7

skipperbent committed:
[BUGFIX] More optimizations

skipperbent committed:
Removed unused RESTServiceItem

skipperbent committed:
More optimisations

skipperbent committed:
[TASK] Refactored structure, naming etc. Made things a little nicer overall.

skipperbent commented:
#15

skipperbent committed:
[REFACTOR] More refactoring..

skipperbent committed:
[TASK] More changes

skipperbent committed:
[REFACTOR] Almost there.. next stop tests

skipperbent committed:
[TASK] Added updated documentation

skipperbent committed:
[TASK] Updated one thing due to the refactor

skipperbent committed:
[TASK] Added documentation page for bonnier-admin (oAuth)

skipperbent committed:
[BUGFIX] Added tests for TRAPP add squashed bugs found

skipperbent committed:
[CLEANUP] Minor cleanup

skipperbent commented:
# Version 2.0

**This release contains breaking changes - please read the changelog and update with caution!**

## Changelog
- Complete refactor of the bonnier-php-sdk to improve performance and avoid unnecessary overflow and naming confusions. 
- Made with extendability in mind - so each class can be easily extended if necessary.
- Added interfaces with events.
- Updated the examples for IndexSearch and added documentation for both OAuth and Trapp (still in progress).
- Added more detailed php-docs-blocks which will be used in the upcoming PHPDocs documentation. 
- Added more PHPUnit tests.
- Bugfixes. 

## Breaking changes in this release
- `IndexDB` classes has been renamed to `IndexSearch` to avoid confusion. All references to the old class names must be renamed to use the new one.
- `getResponse()` method from the ServiceException class has been changed to `getHttpResponse()` and now returns a `HttpResponse` object. Existing code should be changed to `$exception->getHttpResponse()->getResponse()`

## Deprecated methods
- `$content->setData($data)` has been deprecated and will be removed in the near future. Please use `$content->getService()->addPostData($key, $value)` or `$content->getService()->setPostData($array)` instead.

skipperbent commented on src/Bonnier/Admin/OAuth.php:
```diff
@@ -0,0 +1,39 @@
+<?php
```
Ignore this class as there is a seperate pull request for this feature (this one is ooooold)

skipperbent committed:
[BUGFIX] Fixed depricated method

skipperbent committed:
[BUGFIX] Removed unused reference from ServiceLanguage

skipperbent committed:
[BUGFIX] Fixed wrong reference to InvalidArgumentException

alfhen commented on src/Bonnier/HttpRequest.php:
```diff
@@ -0,0 +1,153 @@
+<?php
+namespace Bonnier;
+
+use Doctrine\Instantiator\Exception\InvalidArgumentException;
```
Fix til ikke at være Doctrine

alfhen commented on src/Bonnier/IndexSearch/ServiceRestBase.php:
```diff
@@ -0,0 +1,98 @@
+	 * @throws ServiceException
+	 */
+	protected function onResponseReceived(HttpResponse $originalResponse) {
+		$response = json_decode($originalResponse->getResponse(), TRUE);
```
Fix TRUE til true

skipperbent replied:
Det er fikset for hele projektet - sammen med `FALSE` og `NULL`

alfhen commented on src/Bonnier/IndexSearch/ServiceRestBase.php:
```diff
@@ -0,0 +1,98 @@
+		$data = array_merge($this->request->getPostData(), $data);
+
+		// Add authentication required by index-search
+		$this->request->addHeader('Authorization: Basic ' . base64_encode(sprintf('%s:%s', $this->username, $this->secret)));
```
Lav gerne om til en setBasicAuth($user, $name) på HttpRequest classen.
Så denne linje ville være noget i stil af:

```
$this->request->setBasicAuth($this->username, $this->secret);
```

skipperbent replied:
Den er added i HttpRequest classen.

alfhen commented on src/Bonnier/Trapp/ServiceBase.php:
```diff
@@ -0,0 +1,27 @@
+	// TODO: Add production/staging enviroment when ready (see above)
+
+	protected function getServiceUrl() {
+		return 'http://staging-trapp.whitealbum.dk/api/v1/' . $this->type;
```
Er der en grund til at vi ikke bare adder prod url? 
Den er jo blot http://trapp.whitealbum.dk/api/v1/

Ville det måske også være en idé at abstrahere http/https væk fra getServiceUrl?
Så kunne du have en property der hed secure eller ssl og hvis den var true så brugte man https ellers http.
Så kunne den fx defaulte til https, men så kunne man overskrive det ved at kalde setSsl() eller setSecure().

skipperbent replied:
Vi kan godt lave det i `ServiceBase` (der findes en for både Trapp og IndexSearch) - men kun hvis API'et understøtter en forbindelse hvor man kan fravælge SSL.

Mht. til at smide den i `HttpRequest` så synes jeg ikke det giver mening.

setSSL - ville jo skifte protokol. Men en protokol kan være alt, det kunne også være steam:// eller skype:// - og det er egentlig noget man bør tage stilling til inden man opretter sin `HttpRequest`.

`HttpRequest` former egentlig bare det request til den url du vælger, så hvis den skulle kunne det - burde den tage imod et Url objekt som havde et ``setScheme` eller noget i den duer.

Eksempel:

`$this->service->getUrl()->setScheme('https');`
`$this->service->getUrl()->setHost('lort.dk');`

Lad mig lige høre hvad du synes...

alfhen commented on test/ServiceTranslationTest.php:
```diff
@@ -2,67 +2,88 @@
+		$service->setDevelopment(TRUE);
 
 		// Assert
 		$this->assertNotEquals(TRUE, FALSE);
```
ret capital til true, false

alfhen commented on test/ServiceTranslationTest.php:
```diff
@@ -2,67 +2,88 @@
+		try {
+			$single->update();
+		}catch(\Bonnier\ServiceException $e) {
+			echo sprintf('Error: %s', print_r($e->getHttpResponse()->getResponse(), TRUE));
```
ret capital til lille true

alfhen commented on test/ServiceTranslationTest.php:
```diff
@@ -2,67 +2,88 @@
 
-		$service = new \Bonnier\Trapp\ServiceTranslation('username', '');
+		$service = new \Bonnier\Trapp\ServiceTranslation('Translation', '6277FFAA5D43DEBAF11B62AEB25FB9B5');
+		$service->setDevelopment(TRUE);
```
ret capital til lille true

alfhen commented on src/Bonnier/Trapp/ServiceTranslation.php:
```diff
@@ -1,56 +1,58 @@
-		parent::__construct($username, $secret, self::TYPE);
-		$this->postJson = TRUE;
+		parent::__construct(new ServiceBase($username, $secret, self::TYPE));
+		$this->service->getRequest()->setPostJson(TRUE);
```
ret capital til lille true

alfhen commented on test/ServiceTranslationTest.php:
```diff
@@ -2,67 +2,88 @@
+	public function testDelete() {
 
+		$service = new \Bonnier\Trapp\ServiceTranslation('Translation', '6277FFAA5D43DEBAF11B62AEB25FB9B5');
+		$service->setDevelopment(TRUE);
```
ret capital til lille true

alfhen commented:
Skide godt arbejde! husk at vi skal tagge den nuværende version inden at vi merger disse.

skipperbent committed:
[CLEANUP] Fixed stuff suggested by Alf

- Renamed uppercase constants to lowercase to follow PSR-2.
- Added setBasicAuth($username, $password) method to HttpRequest class.

skipperbent committed:
[TASK] Made ServiceRestBase use $this->request->addBasicAuth($username, $password)

skipperbent commented:
@alfhen 

Tak for reviewet - beklager det blev sådan et kæmpe pull request.
Jeg har svaret på et par af dine kommentarer og fikset de fleste ting i de to forrige commits.

skipperbent committed:
[FEATURE] Fixed headers setHeader instead of addHeader

skipperbent committed:
[BUGFIX] Squashed alot of bugs and restructured some of the code
